src/hyperlight_host/src/hypervisor/whp.rs

```diff
@@ -0,0 +1,410 @@
+
+        unsafe {
+            WHvRunVirtualProcessor(
+                self.partition,
+                0,
+                &mut exit_context as *mut _ as *mut c_void,
+                std::mem::size_of::<WHV_RUN_VP_EXIT_CONTEXT>() as u32,
+            )?;
+        }
+
+        let result = match exit_context.ExitReason {
+            WHvRunVpExitReasonX64IoPortAccess => unsafe {
+                let instruction_length = exit_context.VpContext._bitfield & 0xF;
+                let rip = exit_context.VpContext.Rip + instruction_length as u64;
+                self.set_registers(&[(WHvX64RegisterRip, WHV_REGISTER_VALUE { Reg64: rip })])?;
+                HyperlightExit::IoOut(
+                    exit_context.Anonymous.IoPortAccess.PortNumber,
+                    exit_context
+                        .Anonymous
+                        .IoPortAccess
+                        .Rax
+                        .to_le_bytes()
+                        .to_vec(),
+                )
+            },
+            WHvRunVpExitReasonX64Halt => HyperlightExit::Halt(),
+            WHvRunVpExitReasonMemoryAccess => {
+                let gpa = unsafe { exit_context.Anonymous.MemoryAccess.Gpa };
+                let access_info = unsafe {
+                    WHV_MEMORY_ACCESS_TYPE(
+                        // 2 first bits are the access type, see https://learn.microsoft.com/en-us/virtualization/api/hypervisor-platform/funcs/memoryaccess#syntax
+                        (exit_context.Anonymous.MemoryAccess.AccessInfo.AsUINT32 & 0b11) as i32,
+                    )
+                };
+                let access_info = MemoryRegionFlags::try_from(access_info)?;
+                match access_info {
+                    MemoryRegionFlags::READ => HyperlightExit::MmioRead(gpa),
+                    MemoryRegionFlags::WRITE => HyperlightExit::MmioWrite(gpa),
+                    _ => HyperlightExit::Unknown("Unknown memory access type".to_string()),
+                }
+            }
+            //  Execution was cancelled by the host.
+            //  This will happen when guest code runs for too long
+            WHvRunVpExitReasonCanceled => HyperlightExit::Cancelled(),
+            WHV_RUN_VP_EXIT_REASON(_) => HyperlightExit::Unknown(format!(
+                "Unknown exit reason '{}'",
+                exit_context.ExitReason.0
+            )),
+        };
+        Ok(result)
+    }
+
+    fn get_partition_handle(&self) -> WHV_PARTITION_HANDLE {
+        self.partition
+    }
+}
+
+impl Drop for WhpVm {
+    fn drop(&mut self) {
+        if let Err(e) = unsafe { WHvDeletePartition(self.partition) } {
+            log::error!("Failed to delete partition: {}", e);
+        }
+    }
+}
+
+unsafe fn try_load_whv_map_gpa_range2() -> Result<WHvMapGpaRange2Func> {
+    let library = unsafe {
+        LoadLibraryExA(
+            s!("winhvplatform.dll"),
+            None,
+            LOAD_LIBRARY_SEARCH_DEFAULT_DIRS,
+        )
```

Comment:
Don't think we need this anymore now that in-process is gone. No?

Reply:
No I believe this is different. Here we're checking whether WHvMapGpaRange2 exists.

Reply:
Yes , this is a crappy way on getting a relevant error if running on Windows 10